src/Senders/Jaeger.Senders.Grpc/GrpcSender.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Jaeger.ApiV2;
+using Jaeger.Exceptions;
+using Jaeger.Senders.Grpc.Protocols;
+using GrpcProcess = Jaeger.ApiV2.Process;
+using GrpcSpan = Jaeger.ApiV2.Span;
+
+namespace Jaeger.Senders.Grpc
+{
+    /// <summary>
+    /// GrpcSender provides an implementation to transport spans over HTTP using GRPC.
+    /// </summary>
+    public class GrpcSender : ISender
+    {
+        public const string DefaultCollectorGrpcTarget = "localhost:14250";
+        public const int MaxPacketSize = 65000;
+
+        private readonly Channel _channel;
+        private readonly CollectorService.CollectorServiceClient _client;
+        private readonly int _maxSpanBytes;
+
+        private readonly List<GrpcSpan> _spanBuffer = new List<GrpcSpan>();
+        private GrpcProcess _process;
+        private int _processBytesSize;
+        private int _byteBufferSize;
+
+        /// <summary>
+        /// This constructor expects Jaeger collector running on <see cref="DefaultCollectorGrpcTarget"/> without credentials.
+        /// </summary>
+        public GrpcSender()
+            : this(DefaultCollectorGrpcTarget, ChannelCredentials.Insecure, 0)
+        {
+        }
+
+        /// <param name="target">If empty it will use <see cref="DefaultCollectorGrpcTarget"/>.</param>
+        /// <param name="credentials">If empty it will use <see cref="ChannelCredentials.Insecure"/>.</param>
+        /// <param name="maxPacketSize">If 0 it will use <see cref="MaxPacketSize"/>.</param>
+        public GrpcSender(string target, ChannelCredentials credentials, int maxPacketSize)
+        {
+            if (string.IsNullOrEmpty(target))
+            {
+                target = DefaultCollectorGrpcTarget;
+            }
+
+            if (credentials == null)
+            {
+                credentials = ChannelCredentials.Insecure;
+            }
+
+            if (maxPacketSize == 0)
+            {
+                maxPacketSize = MaxPacketSize;
+            }
+
+            _channel = new Channel(target, credentials);
+            _client = new CollectorService.CollectorServiceClient(_channel);
+            _maxSpanBytes = maxPacketSize;
+        }
+
+        public async Task<int> AppendAsync(Span span, CancellationToken cancellationToken)
```

Comment:
What does the use of async signify? Is this class supposed to provide thread safety for its internal state, like _spanBuffer?

Reply:
async only allows the usage of the await keyword in the method. It brings the compiler in a mode where it generates code returning a Task (promise) instead of the real value. Basically splitting up the method at each await into a asynchronous statemachine.

It has no meaning about safety :)

Reply:
What I mean was whether the use of async might cause a number of parallel executions of this method.

Reply:
Ah yes, it‘s possible that the method is called in parallel, but since one sender is usual only used by one RemoteReporter, this should still be called in sequence and async then is only for interrupt handling.

Reply:
> and async then is only for interrupt handling.

That's what concerns me: RemoteReporter calls Append in one thread, but `await` defers the execution and might continue after the callback on another thread (but I may be misunderstanding the threading behavior in .NET with async).